src/lib/content-indexing.js

```diff
@@ -0,0 +1,98 @@
+/**
+ * @fileoverview Interaction with the Content Indexing API.
+ *
+ * See https://web.dev/content-indexing-api/
+ */
+
+const CACHE_NAME = 'webdev-html-cache-v1';
+const PREFERRED_ICON_SIZE = 192;
+
+const DEFAULT_ICON = {
+  sizes: '128x128',
+  src: '/images/icon-128x128.png',
+  type: 'image/png',
+};
+
+function getIconFromImageSrc(imgSrc) {
+  if (!imgSrc) {
+    return DEFAULT_ICON;
+  }
+
+  const url = new URL(imgSrc, location.href);
+
+  // The image manipulation will only work if this icon is served by Imgix.
+  // If it's not, just use the default icon.
+  if (url.origin !== 'https://webdev.imgix.net') {
+    return DEFAULT_ICON;
+  }
+
+  url.searchParams.set('w', PREFERRED_ICON_SIZE);
+  url.searchParams.set('h', PREFERRED_ICON_SIZE);
+  url.searchParams.set('fit', 'crop');
+  url.searchParams.set('fm', 'webp');
+
+  return {
+    sizes: `${PREFERRED_ICON_SIZE}x${PREFERRED_ICON_SIZE}`,
+    src: url.href,
+    type: 'image/webp',
+  };
+}
+
+// This method syncs the currently cached media with the Content Indexing API
+// (on browsers that support it). The Cache Storage is the source of truth.
+export async function syncContentIndex() {
```

Comment:
Should this also have an "insert" method which is run on SPA navigation? Right now, the index is only updated when I load the site anew.

e.g.:

- I load https://web.dev/, the SW is installed, content is synced (almost nothing)
- I navigate to a few pages
- Content isn't synced

- I load https://web.dev a day later, now those pages get synced

Reply:
It could, yeah. That's how I initially had it implemented, but I moved to a one-off sync at startup in the name of... efficiency, I guess?

I can refactor to do on-the-fly indexing in addition to the one-off startup sync, and I guess if that looks like it's too heavyweight in the real world, we can adjust.

Reply:
See my follow-up comment about where to run that on-the-fly indexing code.